### ISSUE TITLE
Questions about the path constraint for the state after a "CMP/TEST" instruction

### ISSUE THREAD

**Issue by tbbatbb** [angr/angr#2913]
Hi guys, I have a question about the path constraint (the output of calling `state.solver.simplify()`) of the state, which is a branch of a `JS` instruction in `x86`, with a prior instruction `CMP` and `TEST`.

I have to binary files, which are obtained after the compilation of the same source code, with different optimization options, let's say `b1` and `b2`. `b1` is compiled with `-O0`, and `b2` is with '-O2`. The source code will be attached at the last of the post.

As shown in the following picture, the block (or a `state` in `angr`) `A` in binary `b1` is a branch of `JS` instruction, which comes after a `CMP`. 
![image](https://user-images.githubusercontent.com/20682299/133431705-15e6dfb4-663e-4994-a7b5-562a6619c2ba.png)

When it comes to binary `b2`, the instruction `CMP` turns to `TEST`.
![image](https://user-images.githubusercontent.com/20682299/133432306-930a5af9-6d23-4496-b1a8-2c5e666f5665.png)

I use `IDA Pro 7.0` to do the decompilation, and got almost same code, at least at semantic level.
Decompilation result for `b1` with `IDA Pro`:
![image](https://user-images.githubusercontent.com/20682299/133432832-67f3f2eb-289a-4f20-880b-f23d4cfb8631.png)
Decompilation result for `b2` with `IDA Pro`:
![image](https://user-images.githubusercontent.com/20682299/133432967-f3061bf7-9031-47aa-84be-d9455dd921e2.png)

Both are explained as `cmptno >= 0`.

**However, after the symbolic execution with `angr`, I got different output of `state.solver.simplify()`**. For `b1`, the output is:
![image](https://user-images.githubusercontent.com/20682299/133434623-be35d3ea-69aa-4719-ac84-e1e56d80024d.png)
For `b2`, it's:
![image](https://user-images.githubusercontent.com/20682299/133434781-556b9a39-2730-4140-b395-708c6bee8ccd.png)

It's wired for me, I got two different constraints. I don't know whether it's a bug or something, but it does confused me, and made me got into some trouble 🤣 .

Is there any way to make them same?

**Comment by rhelmot** [comment]
@ltfish would it be appropriate to add this optimization to claripy? Could these sort of simplifications be exported from the decompiler and used standalone?

**Comment by ltfish** [comment]
Decompiler’s optimizations work on AIL instead of claripy constraints.

We can have a simplification rule in claripy’s AST simplifier to normalize them to the same constraint. I feel one of them should be faster for Z3 to process than the other.

**Comment by tbbatbb** [comment]
Well, in addition, name of variable `cmptno` appears in constraint because that I simulated the arguments of the function with several `BVS` like `BVS("cmptno", 32, uninitialized=True)`.

**Comment by tbbatbb** [comment]
> Decompiler’s optimizations work on AIL instead of claripy constraints.
> 
> We can have a simplification rule in claripy’s AST simplifier to normalize them to the same constraint. I feel one of them should be faster for Z3 to process than the other.

Ok, I know what happened now. It's my fault.
Constraint in `b2` `cmptno[31:31]==0` checks the `31th` bit in variable `cmptno`. As `cmptno` is treated as a signed integer, so the bit is treated as `sign bit`, and the constraint also makes sure that `cmptno` is positive. But the two representations of the constraints with same semantic meaning do cofused me 🤣 . 

Looking forward to the addition simplification rule! Thanks a lot!!!!

**Comment by tbbatbb** [comment]
OK, maybe one simplification rule is needed. 

When compiling following code:
```c
jas_matrix_t *jas_matrix_create(int numrows, int numcols)
{
	jas_matrix_t *matrix;
	int i;

	if (numrows < 0 || numcols < 0) {
		return 0;
	}

	// several lines of unrelated code

	matrix->maxrows_ = numrows;

        // ...

	if (matrix->maxrows_ > 0) {
		/*
                    ##########################  
		    HERE LOCATES THE TARGET BLOCK
                    ##########################
                */
	}
        // ...
}
```
compiled the code with `-O1` `-O2`, then obtained binaries `b1` and `b2` respectively.

However, the output of `state.solver.simplify()` of the **TARGET BLOCK** on `b1` and `b2` is different.
On `b1`, it's:
![image](https://user-images.githubusercontent.com/20682299/133754580-289d4cbd-39c7-4606-a322-6bba4779c4c4.png)
while on `b2`, it's:
![image](https://user-images.githubusercontent.com/20682299/133754745-d59f3d12-873b-477b-86bf-0c84fd869b19.png)

In short words, the constraints on `b1` is `numrows>0`, and it's `numrows>=0, numrows!=0` on 'b2`.

The case is a little bit different than the above one, and is it possible to make the two constraints identical by adding an additional simplification rule?

**Comment by tbbatbb** [comment]
Hello, is there any progress on this issue? I've met several new cases, like
`numcols[0x1f:0x1f] == 0` V.S. `numcols[0x1f:0x1f] != 0x1`
`numrows>I0x0` V.S. `numrows != I0x0&&numrows[0x1f:0x1f] == I0x0`
`I0x0!=cnt[I0x1F:I0x1F]` V.S. `I0x0>cnt`
`I0x0==M0[I0xA:I0xA]&&I0x0==M0[I0xB:I0xB]` V.S. `I0x0==M0[I0xB:I0xA]`
and many others. They are same in simantic level. 

Is there any short-term plan for this improvement?  😄
